Initial version of README for fixtures (#136)
zkworks-xyz/noir-ethereum-history-api: ethereum_history_api/oracles/README.md

## Noir Ethereum history api oracles

TypeScript oracle server that provides data for circuits.

> ### Preparing fixtures

Reviewer: ```suggestion
### Test fixtures
```

This section will contain instructions about fixture structure, configuration as well as how to prepare them, so the header should be general enough to acomodate that

> Fixtures are data used for testing. They are downloaded and kept as local data. There are two types of fixtures: noir and js. They are pepared in different format and are used in different tests.

Reviewer 2: I don't think this paragraph is needed

Reviewer: What are fixtures is a general knowledge. The fact that we have two types is important, but `different format and are used in different tests` doesn't give any information.

Maybe something like:
### Noir Fixtures
Noir fixtures are generated as noir modules and consumed in Noir tests
### JavaScript Fixtures
JavaScript fixtures are JSON files which cache JSON RPC request/responses pairs
JavaScript fixtures are consumed either directly in unit tests or through a Mock client in integration tests

> To generate data for noir tests run:

Reviewer 2: ```suggestion
To generate data for Noir tests run:
```

```sh
yarn prepare-noir-fixtures
```

> This command will delete current `circuits/lib/src/fixtures` catalog and it will create a new one with new data prepared inside it. Data will be prepared in noir format.

Reviewer 2: Maybe delete this paragraph and insert info about regenerating the folder in the paragraph after prepare-js-fixtures?

Reviewer 2: Or maybe let's not mention that data is being deleted and generated from scratch at all?

> To generate data for js tests run:

Reviewer 2: ```suggestion
To generate data for JavaScript tests run:
```

```sh
yarn prepare-js-fixtures
```

> To configure what data shoud be prepared, use `oracles/src/fixtures/config.ts`. To find a certain Account enter block number and account address. To select Storage from that account additionaly enter storage key. It should be written in the format as shown below:

Reviewer 2: Maybe: 
- "to configure the data that will be prepared, manipulate ... file."
- "to add a certain account ..."
- "to additionally add storage slots enter chosen storage keys"
- "New configuration data needs to adhere to the format below:"
- "Note: storageKeys field can be omitted if generating storage slots is not needed"

```
name_of_account: {
    blockNumber : number_of_block,
    address : address_of_account,
    storageKeys : [storage_keys]
}

> ```

Reviewer: Let's describe that a fixture represents some interesting system state and give examples.
Crypto punk account state at london block
USDC Circle balance as paris block
After this explanation - writing just the type will be obvious:
```ts
interface Fixture {
  blockNumber: bigint;
  address: Address;
  storageKeys?: Hex[];
}

```

# Noir Ethereum history api oracles

TypeScript oracle server that provides data for circuits.

> There are three types of oracles:

Reviewer: ```suggestion
There are three oracles:
```

> - Header Oracle - gets only one argument that is block number
> - Account Oracle - gets two arguments: block number and account address
> - Proof Oracle - gets three arguments: block number, account address and storage key

Reviewer: ```suggestion
- Header Oracle - fetches Ethereum block header
- Account Oracle - fetches Ethereum account state with state proof
- Proof Oracle - fetches Ethereum storage value with both storage & state proof
```

Below are headers for respective functions in Noir:

```rust
fn get_header(block_no: Field) -> (BlockHeaderPartial, BlockHeaderRlp);
fn get_account(block_no: Field, address: Address) -> AccountWithStateProof;
fn get_proof(block_no: Field, address: Address, storage_key: Bytes32) -> StateAndStorageProof;
```

## Starting oracle server

> Oracle server uses connection to Ethereum node. Hence, before running oracle server, first create a `.env` file with a url to Ethereum node. Use template `.env.example` and replace `ETHEREUM_JSON_RPC_API_URL` with your url to rpc server, then the file should be saved as `.env`.

Reviewer: ```suggestion
To run Oracle server, you'll need RPC access to an ETH node.
Set it as an environment variable in `.env` file.
`ETHEREUM_JSON_RPC_API_URL=<url>`
You can copy `.env.example` file.
```

To start oracle server run:

```sh
yarn oracle-server
```

> Following command allows you to use server without restarting it after every change. Server automatically uses new configuration after file change is saved.

Reviewer: ```suggestion
You can also run the server in watch mode:
```

```sh

> yarn oracle-server watch

Reviewer: ```suggestion
yarn oracle-server:watch
```
